### PR TITLE
[IMP] base: sequence, date and time format preview

### PR DESCRIPTION
**Pull request by nla-odoo**
Current behavior before PR:
- when the user makes any change in the sequence they have to do the process that generates the new sequence and checks it
- when the user changes the date time preview or separator for particular the language, they have to go to the SO, Invoice, etc. to check the real value for the format and if the value is not proper they have to change the format again.

Desired behavior after PR is merged:
- users can see the what is a new sequence so they can make changes if needed
- the user can see the preview just beside the format to see if the given format is proper or not

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: odoo/addons/base/models/ir_sequence.py [hunk 1]
```diff
@@ -325,7 +339,7 @@ def _get_number_next_actual(self):
             if seq.sequence_id.implementation != 'standard':
                 seq.number_next_actual = seq.number_next
             else:
-                seq_id = "%03d_%03d" % (seq.sequence_id.id, seq.id)
+                seq_id = "%03d_%03d" % (seq.sequence_id._origin.id, seq._origin.id)
```
Comment (awa-odoo) [inline odoo/addons/base/models/ir_sequence.py:342]: Do we need this here but not in the `_get_number_next_actual` of `ir.sequence` ?
As there is also an id check?

File: odoo/addons/base/models/ir_sequence.py [hunk 2]
```diff
@@ -196,6 +197,19 @@ def write(self, values):
         self.flush_model(values.keys())
         return res
 
+    @api.depends('use_date_range', 'date_range_ids', 'number_next_actual', 'date_range_ids.date_from', 'date_range_ids.date_to', 'date_range_ids.number_next_actual')
+    def _compute_sequence_preview(self):
```
Comment (awa-odoo) [inline odoo/addons/base/models/ir_sequence.py:201]: IMO this should depend on "prefix" and "suffix", that way if you modify those fields the preview is correctly updated.

File: odoo/addons/base/tests/test_ir_sequence.py [hunk 3]
```diff
@@ -175,10 +176,40 @@ def test_ir_sequence_create_no_gap(self):
                 n = env['ir.sequence'].next_by_code('test_sequence_type_6')
                 self.assertEqual(n, str(i))
 
+    def test_ir_sequence_preview(self):
+        with environment() as env:
+            seq = env['ir.sequence'].create({
+                'code': 'test_sequence_type_7',
+                'name': 'Test sequence',
+                'number_next_actual': '123',
+                'prefix': 'Test-'
```
Comment (awa-odoo) [inline odoo/addons/base/tests/test_ir_sequence.py:185]: Could add some suffix as well to check that we correctly build everything.

File: odoo/addons/base/views/ir_sequence_views.xml [hunk 4]
```diff
@@ -10,6 +10,7 @@
                       <group>
                         <field name="name"/>
                         <field name="implementation"/>
+                        <field name="sequence_preview" class="text-muted"/>
```
Comment (awa-odoo) [inline odoo/addons/base/views/ir_sequence_views.xml:13]: Not sure why it's text-muted, makes it look strange in comparison to the rest of the form.

File: odoo/addons/base/views/res_lang_views.xml [hunk 5]
```diff
@@ -68,6 +68,15 @@
                                 <field name="time_format"/>
                                 <field name="week_start"/>
                             </group>
+                            <div class='mt-4 border border-3 p-3 text-muted col-lg-4'>
+                                <p class="text-center">Preview</p>
+                                <group>
+                                    <field name="date_format_preview" widget="lang_date" attrs="{'invisible': [('date_format', 'in', ['', False])]}"/>
```
Comment (awa-odoo) [inline odoo/addons/base/views/res_lang_views.xml:74]: Would it be possible to implement it the same way as the other ones, using a python computed field (Char)?

Having a new JS widget seems very heavy.

File: odoo/addons/base/models/res_lang.py [hunk 6]
```diff
@@ -94,6 +98,18 @@ def _check_grouping(self):
...
                 raise ValidationError(warning)
 
+    @api.depends('decimal_point')
+    def _compute_decimal_point_preview(self):
+        self.decimal_point_preview = False
+        for lang in self.filtered('decimal_point'):
+            lang.decimal_point_preview = '99' + lang.decimal_point + '00'
+
+    @api.depends('thousands_sep', 'grouping')
+    def _compute_thounsand_separator_preview(self):
```
Comment (awa-odoo) [inline odoo/addons/base/models/res_lang.py:108]: Typo: extra "n" in "thounsand"

File: odoo/addons/base/models/res_lang.py [hunk 7]
```diff
@@ -46,7 +48,9 @@ class Lang(models.Model):
              "[1,2,-1] will represent it to be 106,50,0;[3] will represent it as 106,500. "
              "Provided ',' as the thousand separator in each case.")
     decimal_point = fields.Char(string='Decimal Separator', required=True, default='.', trim=False)
+    decimal_point_preview = fields.Char(string="Decimal separator preview", compute="_compute_decimal_point_preview")
     thousands_sep = fields.Char(string='Thousands Separator', default=',', trim=False)
+    thousand_separator_preview = fields.Char(string="Thousands separator preview", compute='_compute_thounsand_separator_preview')
```
Comment (awa-odoo) [inline odoo/addons/base/models/res_lang.py:53]: Would name it "thousands_sep_preview" to match the other field.

File: odoo/addons/base/models/res_lang.py [hunk 8]
```diff
@@ -46,7 +48,9 @@ class Lang(models.Model):
              "[1,2,-1] will represent it to be 106,50,0;[3] will represent it as 106,500. "
              "Provided ',' as the thousand separator in each case.")
     decimal_point = fields.Char(string='Decimal Separator', required=True, default='.', trim=False)
+    decimal_point_preview = fields.Char(string="Decimal separator preview", compute="_compute_decimal_point_preview")
```
Comment (awa-odoo) [inline odoo/addons/base/models/res_lang.py:51]: We usually capitalize everything in labels, so "Decimal Separator Preview".

(to apply to other new fields).

**Review by awa-odoo — COMMENTED** [review]
Hello @nla-odoo 

Neat little feature.
I added a few comments to try to clean / simplify.

Thanks & cheers!

File: odoo/addons/base/views/res_lang_views.xml (hunk 5 above)
Comment (nla-odoo) [reply odoo/addons/base/views/res_lang_views.xml:74]: Hello @awa-odoo 

The reason is we want to show the translation based on the current record language not from the user language.
and we don't want to store all the translations for the formats
that can be anything, hence it is suitable to do in JS. so, the user can use any format and the DateTime formater will handle it automatically.

Thanks

Comment (awa-odoo) [reply odoo/addons/base/views/res_lang_views.xml:74]: Hello @nla-odoo 

Still not convinced on this one.
You could use the methods in tools to format_date and format_time based on a specific given format:
https://github.com/odoo/odoo/blob/master/odoo/tools/misc.py#L1359

So you would depend on the date_format and time_format fields in a compute method and change your preview accordingly.
Maybe I'm missing your point though, I don't really understand the part "and we don't want to store all the translations for the formats".

From what I understand, we need to use a format that would be supported by the backend anyway, as this is going to be used in the backend too once the res.lang is configured. 

Thank you for investigating & cheers!

File: odoo/addons/base/models/res_lang.py [hunk 9]
```diff
@@ -94,6 +99,34 @@ def _check_grouping(self):
             except Exception:
                 raise ValidationError(warning)
 
+    @api.depends('date_format')
+    def _compute_date_format_preview(self):
```
Comment (awa-odoo) [inline odoo/addons/base/models/res_lang.py:103]: I would mention in the docstring that we use direct babel calls instead of tooling methods because the language may not be active yet.
(Same for the time formatting).

File: odoo/addons/base/tests/test_res_lang.py [hunk 10]
```diff
@@ -40,3 +42,63 @@ def test_00_intersperse(self):
         assert intersperse("abc12", [3], '.') == ('abc12', 0)
         assert intersperse("abc12", [2], '.') == ('abc12', 0)
         assert intersperse("abc12", [1], '.') == ('abc1.2', 1)
+
+    def test_res_lang_decimal_point_preview(self):
+        lang_0 = self.env['res.lang'].create({
```
Comment (awa-odoo) [inline odoo/addons/base/tests/test_res_lang.py:47]: You could add a setupClass method to create a test_lang record.

That way we gain one creation as it's used in multiple tests.

File: odoo/addons/base/tests/test_res_lang.py [hunk 11]
```diff
@@ -40,3 +42,63 @@ def test_00_intersperse(self):
...
+
+        lang_0.write({'decimal_point': ','})
+
+        self.assertEqual(lang_0.decimal_point_preview, '99,00')
+
+    def test_res_lang_thousand_sep_preview(self):
+        lang_0 = self.env['res.lang'].create({
+            'name': 'Test Lang',
+            'iso_code': 'TL',
+            'code': 'TL',
```
Comment (awa-odoo) [inline odoo/addons/base/tests/test_res_lang.py:62]: To avoid any possible collisions I would go for "test_lang" for both of those.

File: odoo/addons/base/tests/test_res_lang.py [hunk 12]
```diff
@@ -40,3 +42,63 @@ def test_00_intersperse(self):
...
+
+        self.assertEqual(lang_0.thousand_sep_preview, '999.999.999')
+
+    def test_res_lang_date_format_preview(self):
+        class MockedDatetime(datetime):
+            @classmethod
+            def now(cls):
+                return datetime(2020, 1, 1, 16, 0, 0)
+
+        with patch('odoo.fields.Datetime', MockedDatetime):
```
Comment (awa-odoo) [inline odoo/addons/base/tests/test_res_lang.py:78]: Same here, could you use freeze_time?

You should be able to use it like this:

```
@freeze_time("2020-01-01 16:00:00")
def test_res_lang_date_format_preview(self):
    ...
```

File: odoo/addons/base/tests/test_res_lang.py [hunk 13]
```diff
@@ -40,3 +42,63 @@ def test_00_intersperse(self):
...
+            self.assertEqual(lang_0.date_format_preview, '01/01/2020')
+            self.assertEqual(lang_1.date_format_preview, '01 يناير, 2020')
+
+            (lang_0 | lang_1).write({'date_format': '%A-%B-%Y'})
+
+            self.assertEqual(lang_0.date_format_preview, 'Wednesday-January-2020')
+            self.assertEqual(lang_1.date_format_preview, 'الأربعاء-يناير-2020')
+
+    def test_res_lang_time_format_preview(self):
+        class MockedDatetime(datetime):
```
Comment (awa-odoo) [inline odoo/addons/base/tests/test_res_lang.py:90]: Same remark about freeze_time.

**Review by awa-odoo — COMMENTED** [review]
Hello @nla-odoo 

New implementation is looking good.
A few remarks left to try to polish, especially in the tests.

After that we should be good to go.

Thanks for updating & cheers!

File: odoo/addons/base/models/res_lang.py (hunk 9 above)
Comment (nla-odoo) [reply odoo/addons/base/models/res_lang.py:103]: Hello @awa-odoo 

Changes are done.

Thanks

File: odoo/addons/base/tests/test_res_lang.py (hunk 11 above)
Comment (awa-odoo) [reply odoo/addons/base/tests/test_res_lang.py]: Does not seem addressed? Maybe you did not push the latest version?

File: odoo/addons/base/models/res_lang.py [hunk 14]
```diff
@@ -94,6 +99,40 @@ def _check_grouping(self):
...
+        """ Sometimes language may not be activated and the format DateTime would not work as expected.
+        hence, we are directly using Bebel to parse the required language to format the date.
```
Comment (awa-odoo) [inline odoo/addons/base/models/res_lang.py:105]: Bebel -> babel ^^

(same for the other comment)

File: odoo/addons/base/tests/test_ir_sequence.py [hunk 15]
```diff
@@ -175,10 +176,41 @@ def test_ir_sequence_create_no_gap(self):
...
+                }), (0, 0, {
+                    'date_from': odoo.fields.Date.today() + timedelta(days=-5),
+                    'date_to': odoo.fields.Date.today() + timedelta(days=15),
+                    'number_next_actual': 12345
+                })]
+            })
+            self.assertEqual(seq.sequence_preview, 'Test-12345-Odoo')
+
+            seq.date_range_ids[0].write({
+                'date_from': odoo.fields.Date.today() + timedelta(days=-5),
```
[the remark this answers is not in the document]
Comment (nla-odoo) [reply odoo/addons/base/tests/test_ir_sequence.py]: Hello @awa-odoo ,

Sure, I will do it.
But here the case I have handled is, if we have overlapping date ranges, it should consider the latest created range.

Thanks

[the remark this answers is not in the document]
Comment (awa-odoo) [reply odoo/addons/base/tests/test_ir_sequence.py]: Hello @nla-odoo 

I see what you mean, I consider that kind of an implementation detail though.
It's interesting to test maybe in another case (although probably overkill), but for the base testing it feels strange to test a side case.

Thanks & cheers!

[the remark this answers is not in the document]
Comment (nla-odoo) [reply odoo/addons/base/tests/test_ir_sequence.py]: Hello @awa-odoo 

Changes are done.

Thanks

**Comment by Julien00859** [comment]
Could any of you too set a profile picture? It took me too much time to realize there was two people involved x)))

up @odoo/rd-framework-py

File: odoo/addons/base/tests/test_res_lang.py [hunk 16]
```diff
@@ -40,3 +51,41 @@ def test_00_intersperse(self):
...
+        self.assertEqual(self.lang_0.thousand_sep_preview, '999.999.999')
+
+    @freeze_time("2020-01-01 16:00:00")
+    def test_res_lang_date_format_preview(self):
+        lang_0, lang_1 = self.env['res.lang'].with_context(active_test=False).search(
+            [('iso_code', 'in', ['en', 'ar'])]
+        )
+
+        self.assertEqual(lang_0.date_format_preview, '01/01/2020')
+        self.assertEqual(lang_1.date_format_preview, '01 يناير, 2020')
```
Comment (VincentSchippefilt) [inline odoo/addons/base/tests/test_res_lang.py:72]: Use any language you want, provided it is based on latin charset. If in the future we need to change this test we (at lease I) can't even type in arabic, let alone verify what you wrote.
Could you change the language in this test ?

File: odoo/addons/base/models/ir_sequence.py [hunk 17]
```diff
@@ -196,6 +197,19 @@ def write(self, values):
         self.flush_model(values.keys())
         return res
 
+    @api.depends('prefix', 'suffix', 'use_date_range', 'date_range_ids', 'date_range_ids.date_from', 'date_range_ids.date_to', 'date_range_ids.number_next_actual', 'number_next_actual')
```
Comment (VincentSchippefilt) [inline odoo/addons/base/models/ir_sequence.py:200]: missing dependency on `paddind`?

**Review by VincentSchippefilt — CHANGES_REQUESTED** [review]
please update your tests and add the missing dependency in ir_sequence

File: odoo/addons/base/models/res_lang.py [hunk 18]
```diff
@@ -111,6 +116,48 @@ def _check_grouping(self):
...
+    def _compute_time_format_preview(self):
+        """ Sometimes language may not be activated and the format DateTime would not work as expected.
+        hence, we are directly using babel to parse the required language to format the date.
+        """
+        self.time_format_preview = False
+        for lang in self.filtered(lambda lang: lang.time_format and lang.code):
+            try:
+                locale = tools.babel_locale_parse(lang.code)
+                time_format = tools.posix_to_ldml(lang.time_format, locale=locale)
+                lang.time_format_preview = '(e.g. "' + babel.dates.format_time(fields.Datetime.now(), time_format, locale=locale) + '")'
```
Comment (std-odoo) [inline odoo/addons/base/models/res_lang.py:155]: same here and above about translation

File: odoo/addons/base/models/res_lang.py [hunk 19]
```diff
@@ -111,6 +116,48 @@ def _check_grouping(self):
...
+        """ Sometimes language may not be activated and the format DateTime would not work as expected.
+        hence, we are directly using babel to parse the required language to format the date.
+        """
+        self.date_format_preview = False
+        for lang in self.filtered(lambda lang: lang.date_format and lang.code):
+            locale = tools.babel_locale_parse(lang.code)
+            try:
+                date_format = tools.posix_to_ldml(lang.date_format, locale=locale)
+                lang.date_format_preview = '(e.g. "' + babel.dates.format_datetime(fields.Datetime.now(), date_format, locale=locale) + '")'
+            except KeyError:
```
Comment (std-odoo) [inline odoo/addons/base/models/res_lang.py:130]: instead of catching the exception, can't we check that `lang.code` is in the list ?
(and then remove the try / except)

same for other methods

File: odoo/addons/base/models/res_lang.py [hunk 20]
```diff
@@ -32,7 +33,9 @@ class Lang(models.Model):
     active = fields.Boolean()
     direction = fields.Selection([('ltr', 'Left-to-Right'), ('rtl', 'Right-to-Left')], required=True, default='ltr')
     date_format = fields.Char(string='Date Format', required=True, default=DEFAULT_DATE_FORMAT)
+    date_format_preview = fields.Char(string='Date format preview', compute="_compute_date_format_preview", readonly=True, store=False)
```
Comment (std-odoo) [inline odoo/addons/base/models/res_lang.py:36]: this is already the default for a computed field

(same below)
```suggestion
    date_format_preview = fields.Char(string='Date format preview', compute="_compute_date_format_preview")

File: odoo/addons/base/tests/test_res_lang.py [hunk 21]
```diff
@@ -40,3 +51,41 @@ def test_00_intersperse(self):
...
+        self.assertEqual(self.lang_0.decimal_point_preview, '(e.g. "99,00")')
+
+    def test_res_lang_thousand_sep_preview(self):
+        self.assertEqual(self.lang_0.thousand_sep_preview, '(e.g. "999,999,999")')
+        self.lang_0.write({'thousands_sep': '.'})
+        self.assertEqual(self.lang_0.thousand_sep_preview, '(e.g. "999.999.999")')
+
+    @freeze_time("2020-01-01 16:00:00")
+    def test_res_lang_date_format_preview(self):
+        lang_0, lang_1 = self.env['res.lang'].with_context(active_test=False).search(
```
Comment (std-odoo) [inline odoo/addons/base/tests/test_res_lang.py:67]: can you order by `iso_code` to make the test more robust ?

File: odoo/addons/base/tests/test_res_lang.py [hunk 22]
```diff
@@ -40,3 +51,41 @@ def test_00_intersperse(self):
...
+        self.assertEqual(lang_1.date_format_preview, '(e.g. "01/01/2020")')
+
+        (lang_0 | lang_1).write({'date_format': '%A-%B-%Y'})
+
+        self.assertEqual(lang_0.date_format_preview, '(e.g. "Wednesday-January-2020")')
+        self.assertEqual(lang_1.date_format_preview, '(e.g. "mercredi-janvier-2020")')
+
+    @freeze_time("2020-01-01 16:00:00")
+    def test_res_lang_time_format_preview(self):
+        lang_0, lang_1 = self.env['res.lang'].with_context(active_test=False).search(
```
Comment (std-odoo) [inline odoo/addons/base/tests/test_res_lang.py:81]: same here

File: odoo/addons/base/tests/test_res_lang.py [hunk 23]
```diff
@@ -40,3 +51,41 @@ def test_00_intersperse(self):
...
+            [('iso_code', 'in', ['en', 'fr'])]
+        )
+
+        self.assertEqual(lang_0.time_format_preview, '(e.g. "16:00:00")')
+        self.assertEqual(lang_1.time_format_preview, '(e.g. "16:00:00")')
+
+        (lang_0 | lang_1).write({'time_format': '%M-%I-%S'})
```
Comment (std-odoo) [inline odoo/addons/base/tests/test_res_lang.py:88]: (same above for other tests)

```suggestion
        (lang_0 | lang_1).time_format = '%M-%I-%S'

File: odoo/addons/base/tests/test_res_lang.py [hunk 24]
```diff
@@ -1,10 +1,21 @@
...
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
+from freezegun import freeze_time
 
 from odoo.tests.common import TransactionCase
 
 class test_res_lang(TransactionCase):
 
+    @classmethod
+    def setUpClass(cls):
+        super(test_res_lang, cls).setUpClass()
```
Comment (std-odoo) [inline odoo/addons/base/tests/test_res_lang.py:11]:
```suggestion
        super().setUpClass()

File: odoo/addons/base/views/ir_sequence_views.xml [hunk 25]
```diff
@@ -26,9 +26,14 @@
                             <field name="use_date_range"/>
                           </group>
                           <group>
-                            <field name="padding"/>
-                            <field name="number_increment"/>
-                            <field name="number_next_actual" string="Next Number" attrs="{'invisible': [('use_date_range', '=', True)]}"/>
+                              <group>
+                                <field name="padding"/>
+                                <field name="number_increment"/>
+                                <field name="number_next_actual" string="Next Number" attrs="{'invisible': [('use_date_range', '=', True)]}"/>
```
Comment (std-odoo) [inline odoo/addons/base/views/ir_sequence_views.xml:32]: can be multi-lined

File: odoo/addons/base/models/ir_sequence.py [hunk 26]
```diff
@@ -196,6 +197,19 @@ def write(self, values):
         self.flush_model(values.keys())
         return res
 
+    @api.depends('prefix', 'suffix', 'padding', 'use_date_range', 'date_range_ids', 'date_range_ids.date_from', 'date_range_ids.date_to', 'date_range_ids.number_next_actual', 'number_next_actual')
```
Comment (std-odoo) [inline odoo/addons/base/models/ir_sequence.py:200]: why does it depends on `padding` ?

File: odoo/addons/base/models/ir_sequence.py [hunk 27]
```diff
@@ -103,7 +103,7 @@ def _get_number_next_actual(self):
             elif seq.implementation != 'standard':
                 seq.number_next_actual = seq.number_next
             else:
-                seq_id = "%03d" % seq.id
+                seq_id = "%03d" % seq._origin.id
```
Comment (std-odoo) [inline odoo/addons/base/models/ir_sequence.py:106]: why this change ?

File: odoo/addons/base/models/ir_sequence.py (hunk 1 above)
Comment (std-odoo) [inline odoo/addons/base/models/ir_sequence.py:342]: why this change ?

**Review by std-odoo — COMMENTED** [review]
Hi :) 

Not sure who to ping here, but I have a few coments / questions

File: odoo/addons/base/models/ir_sequence.py (hunk 26 above)
Comment (prra-odoo) [reply odoo/addons/base/models/ir_sequence.py:200]: We will have to update the preview based on changes in the `padding` field and for that reason, it is added in depends (was added while validation).

Comment (std-odoo) [reply odoo/addons/base/models/ir_sequence.py:202]: but we don't use the padding in the compute, why do we want to do that ?

Comment (prra-odoo) [reply odoo/addons/base/models/ir_sequence.py:202]: the `get_next_char` uses the padding, thus change in the padding field should be reflected in the preview

Comment (std-odoo) [reply odoo/addons/base/models/ir_sequence.py:202]: ah ok, thanks for explanation :)

**Comment by tde-banana-odoo** [comment]
Canceled (see task).
